### ISSUE TITLE
NRE with nullable subselect value in Linq

### ISSUE THREAD

**Issue by NikitaEgorov** [nhibernate/nhibernate-core#2425]
**Version:** NHibernate-5.3.0
**Exception:** 
Object reference not set to an instance of an object.

```c#
   at System.Linq.Enumerable.SelectEnumerableIterator`2.ToList()
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at NHibernate.Linq.ResultTransformer.TransformList(IList collection) in r:\nhibernate-core-master\src\NHibernate\Linq\ResultTransformer.cs:line 36

```
**LINQ Expression**:

```c#
.Select(
    x => new
        {
            x.Date,
            x.Employee.Number,
            CountyNumber = (int?)x.Site.CountyNumber, <=== Pass through null value
            Item = x.Items.Select(zx => new { zx.Value }) <=== Subselect
        })
```

listTransformers[1].ToString()

```c#
inputList => inputList
    .Select(g => new <>f__AnonymousType1`4(
                Date = Convert(g.Key.Items[1], Nullable`1), 
                Number = Convert(g.Key.Items[2], Int32), 
                CountyNumber = Convert(Convert(g.Key.Items[3], Int32), Nullable`1), <======== Int32 shoud be Nullable
                Item = Convert(g.Where(t => Not(ReferenceEquals(t.Items[0], null))).Select(value => new <>f__AnonymousType2`1(BaseValue = Convert(value.Items[1], Nullable`1))).ToList().AsReadOnly(), IEnumerable`1))).ToList()
```

Exception happened only if Select contains  "Subselect" and any value "pass" through null value

**Comment by maca88** [comment]
This issue should be fixed by #2079, where a friendlier exception is thrown when the cast is missing:
https://github.com/nhibernate/nhibernate-core/blob/04e40931af0d812810ef176deb452847154b6ee9/src/NHibernate.Test/Linq/SelectionTests.cs#L1057-L1081

In order to assure that the issue is fixed by the mentioned PR, it would be nice if you provide the whole linq query so that it can be added to the tests.

**Comment by NikitaEgorov** [comment]
I reproduced this issue on master branch nhibernate

**Comment by maca88** [comment]
The mentioned PR is not yet megred into master, you need to build the code from the PR in order to test it.

**Comment by NikitaEgorov** [comment]
```c#
[Test]
public void CanExecuteMethodWithNullObjectAndSubselect()
{
    var list1 = db.Animals.Select(
                                    a => new
                                        {
                                            NullableId = (int?) a.Father.Father.Id,
                                        })
                    .ToList();
    Assert.That(list1, Has.Count.GreaterThan(0));
    Assert.That(list1[0].NullableId, Is.Null);

    var list2 = db.Animals.Select(
                                    a => new
                                        {
                                            Descriptions = a.Children.Select(z => z.Description)
                                        })
                    .ToList();
    Assert.That(list2, Has.Count.GreaterThan(0));
    Assert.That(list2[0].Descriptions, Is.Not.Null);

    var list3 = db.Animals.Select(
                                a => new
                                        {
                                            NullableId = (int?) a.Father.Father.Id,
                                            Descriptions = a.Children.Select(z=>z.Description)
                                        })
                .ToList();
    Assert.That(list3, Has.Count.GreaterThan(0));
    Assert.That(list1[0].NullableId, Is.Null);
    Assert.That(list2[0].Descriptions, Is.Not.Null);
}
```

```c#

NHibernate.Exceptions.GenericADOException : Could not execute query[SQL: SQL not available]
  ----> System.NullReferenceException : Object reference not set to an instance of an object.
   at NHibernate.Impl.SessionImpl.List(IQueryExpression queryExpression, QueryParameters queryParameters, IList results, Object filterConnection) in r:\nhibernate-core-GH2054\src\NHibernate\Impl\SessionImpl.cs:line 560
   at NHibernate.Impl.SessionImpl.List(IQueryExpression queryExpression, QueryParameters queryParameters, IList results) in r:\nhibernate-core-GH2054\src\NHibernate\Impl\SessionImpl.cs:line 520
   at NHibernate.Impl.AbstractSessionImpl.List[T](IQueryExpression query, QueryParameters parameters) in r:\nhibernate-core-GH2054\src\NHibernate\Impl\AbstractSessionImpl.cs:line 150
   at NHibernate.Impl.AbstractQueryImpl2.List[T]() in r:\nhibernate-core-GH2054\src\NHibernate\Impl\AbstractQueryImpl2.cs:line 111
   at NHibernate.Linq.DefaultQueryProvider.ExecuteList[TResult](Expression expression) in r:\nhibernate-core-GH2054\src\NHibernate\Linq\DefaultQueryProvider.cs:line 111
   at NHibernate.Linq.NhQueryable`1.System.Collections.Generic.IEnumerable<T>.GetEnumerator() in r:\nhibernate-core-GH2054\src\NHibernate\Linq\NhQueryable.cs:line 65
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at NHibernate.Test.Linq.SelectionTests.CanExecuteMethodWithNullObjectAndSubselect() in r:\nhibernate-core-GH2054\src\NHibernate.Test\Linq\SelectionTests.cs:line 1167
--NullReferenceException
   at lambda_method(Closure , IGrouping`2 )
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at NHibernate.Linq.ResultTransformer.TransformList(IList collection) in r:\nhibernate-core-GH2054\src\NHibernate\Linq\ResultTransformer.cs:line 36
   at NHibernate.Loader.Hql.QueryLoader.GetResultList(IList results, IResultTransformer resultTransformer) in r:\nhibernate-core-GH2054\src\NHibernate\Loader\Hql\QueryLoader.cs:line 343
   at NHibernate.Loader.Loader.ListIgnoreQueryCache(ISessionImplementor session, QueryParameters queryParameters) in r:\nhibernate-core-GH2054\src\NHibernate\Loader\Loader.cs:line 1837
   at NHibernate.Loader.Loader.List(ISessionImplementor session, QueryParameters queryParameters, ISet`1 querySpaces) in r:\nhibernate-core-GH2054\src\NHibernate\Loader\Loader.cs:line 1827
   at NHibernate.Loader.Hql.QueryLoader.List(ISessionImplementor session, QueryParameters queryParameters) in r:\nhibernate-core-GH2054\src\NHibernate\Loader\Hql\QueryLoader.cs:line 325
   at NHibernate.Hql.Ast.ANTLR.QueryTranslatorImpl.List(ISessionImplementor session, QueryParameters queryParameters) in r:\nhibernate-core-GH2054\src\NHibernate\Hql\Ast\ANTLR\QueryTranslatorImpl.cs:line 131
   at NHibernate.Engine.Query.HQLQueryPlan.PerformList(QueryParameters queryParameters, ISessionImplementor session, IList results) in r:\nhibernate-core-GH2054\src\NHibernate\Engine\Query\HQLQueryPlan.cs:line 115
   at NHibernate.Impl.SessionImpl.List(IQueryExpression queryExpression, QueryParameters queryParameters, IList results, Object filterConnection) in r:\nhibernate-core-GH2054\src\NHibernate\Impl\SessionImpl.cs:line 550
```

**Comment by maca88** [comment]
Thanks for providing the test case, I made a fix in #2428.
